runtime/interpreter/value_string.go

```diff
@@ -108,6 +108,74 @@ func stringFunctionFromCharacters(invocation Invocation) Value {
 	return NewUnmeteredStringValue(builder.String())
 }
 
+var StringTypeJoinDefaultSeparator = NewUnmeteredStringValue(",")
+
+func stringFunctionJoin(invocation Invocation) Value {
+	argument, ok := invocation.Arguments[0].(*ArrayValue)
+	if !ok {
+		panic(errors.NewUnreachableError())
+	}
+
+	inter := invocation.Interpreter
+
+	switch argument.Count() {
+	case 0:
+		return EmptyString
+	case 1:
+		return argument.Get(inter, invocation.LocationRange, 0)
+	}
+
+	var separator *StringValue
+	if len(invocation.Arguments) > 1 {
+		separator, ok = invocation.Arguments[1].(*StringValue)
+		if !ok {
+			panic(errors.NewUnreachableError())
+		}
+	} else {
+		separator = StringTypeJoinDefaultSeparator
+	}
+
+	common.UseMemory(inter,
+		common.MemoryUsage{
+			Kind:   common.MemoryKindStringValue,
+			Amount: 1,
+		},
+	)
+	var builder strings.Builder
+	first := true
+
+	argument.Iterate(inter, func(element Value) (resume bool) {
+		// Add separator
+		if !first {
+			common.UseMemory(inter,
+				common.MemoryUsage{
+					Kind:   common.MemoryKindStringValue,
+					Amount: uint64(len(separator.Str)),
+				},
```

Comment:
Use `NewStringMemoryUsage` instead. Also please use it in `stringFunctionFromCharacters` and `TypeValue.GetMember` while at it.

Reply:
Done, do note that in some places in `stringFunctionJoin` and `stringFunctionFromCharacters`, I had to do a `-1` to account for double counting of empty string.